TFT_LITTLE_VGL_UI for everyone

rhapsodyv opened the pull request:
### Description

Fixes that make TFT_LITTLE_VGL_UI working!!!

I fixed the MKS code, and added a few switches to use the same UI in others screens. I'm currently using it in a Tronxy X5SA stock board and TFT.

I think we can wait the MKS dev, before merge this code. 

### Benefits

More one screen option for Marlin users (not only mks users)

### Related Issues

#18401 
#18410

rhapsodyv committed:
Fixes PR #18071 TFT_LITTLE_VGL_UI broken compilation

thinkyhead committed:
Update draw_error_message.h

rhapsodyv committed:
fix a lot of issues with TFT_LITTLE_VGL_UI. Now its fully working in fsmc dma mode - SPI_GRAPHICAL_TFT not tested, even in a tronxy stock tft!

rhapsodyv committed:
chitu boards with TFT Little VGL UI

rhapsodyv committed:
keep defaults as it was in the mks code

rhapsodyv committed:
correct name

rhapsodyv committed:
correct platform

rhapsodyv committed:
now compliling fine

thinkyhead committed:
Update MarlinCore.cpp

thinkyhead committed:
Fix setting of board name

thinkyhead commented on Marlin/src/lcd/extui/lib/mks_ui/draw_error_message.h:
```diff
@@ -25,6 +25,10 @@
+  #define PGM_P const char *
+#endif
+
 extern void lv_draw_error_message(PGM_P const msg);
```
`PGM_P` specifically refers to strings stored in Flash and defined with `PROGMEM` or `PSTR`. It should not be used for C strings residing in SRAM or other static strings with no `PROGMEM` specifier.

rhapsodyv replied:
I think this code isnt used anywhere... but the original MKS PR doesn't compile without that define. I will try test without this code.
I just didnt remove because the MKS can have a reason for that code...

thinkyhead replied:
We'll have to pass it on to MKS. `PGM_P` is meant to be a clue about the string storage.

rhapsodyv replied:
@makerbase-mks the code above isn't used anywhere...I compiled and run a version without it with apparently no problem. Can I remove or comment it by now? It will have a future usage?

thinkyhead committed:
Update draw_ui.h

thinkyhead committed:
Update draw_ui.h

thinkyhead committed:
Update draw_ui.cpp

thinkyhead committed:
Update draw_ui.cpp

thinkyhead committed:
General cleanup

rhapsodyv commented:
@thinkyhead 

I forgot to say: 

The code loads some .bin files from mks_pic/mks_font folders from SD to SPI Flash. The asset load runs when this folders are in the SD root. The code uses the filenames as key.

But, there're 2 problems: 

1) The code are referencing bin files that aren't in the MKS release files. So, it won't work 100% if the user use the mks supplied mks_* folder. I think they renamed the files, but forgot the code...

2) The lookup is case sensitive. But, the actual code sent by MKS doesn't match the case of their .bin files....

Until we get a MKS response, I have created a folder with all necessary bin files with correct names and case:

https://drive.google.com/file/d/1CSCi-306uCIV7Yz9wasTV3_5iZpTSmIO/view?usp=sharing

I think we need standardise the assets spi storage and lookup/method. But by know, I wanted to keep the original MKS code to avoid more conflicts.

thinkyhead committed:
Strip pic_manager dead code

thinkyhead committed:
FSMC define

thisiskeithb commented:
Let me know once it’s ready for review and I’ll test it on a Robin Nano with TFT35.

thinkyhead committed:
Strip tft_lvgl_configuration

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/18438

thinkyhead committed:
NO_LCD_REINIT not needed

thinkyhead committed:
Apply pins formatting js

thinkyhead committed:
tweaks

thinkyhead committed:
Update env and SoftwareSerialM

thinkyhead commented:
I've done a bunch of cleanup and started stripping out the dead code to get it down to the minimum, in case someone wants to work on the code in future.

I saw that the `SoftwareSerialM` library is in the registry, so removed the long URL for `master.zip`.

rhapsodyv commented:
Ok, I will pul changes and re-test everything.

thisiskeithb commented:
I flashed a Robin Nano 1.2 + TFT35 combo and it stays stuck at "PIC UPDATING..." until power cycled. It boots/the backlight turns on, but but then nothing is displayed on the screen.

Clicking around to see if icons were loaded, but possibly not displaying correctly like I was seeing before this PR doesn't work. It's just... not working at all.

rhapsodyv commented:
> I flashed a Robin Nano 1.2 + TFT35 combo (using the updated icons from above) and it stays stuck at "PIC UPDATING..." until power cycled. It boots/the backlight turns on, but but then nothing is displayed on the screen.
> 
> Tapping around to see if icons were loaded, but possibly not displaying correctly like I was seeing before this PR doesn't work. It's just... not working at all.

Which is your SPI Flash?

thisiskeithb commented:
I'm not sure. Whatever is stock on the Robin Nano board/screen. It's an `STM32F103VET6`-based board.

rhapsodyv commented:
I think I know. The font load is broken. But I just dont changed it. I just disabled for my board. Disable it in your config file and test again, please.

```cpp
  #define HAS_SPI_FLASH_FONT 0
  #define HAS_GCODE_PREVIEW 0
```

rhapsodyv commented:
I could not test 3 things:

1) Font bin load (the original code is broken and I dont have spi flash enough for that too, so I just disabled with a define and kept the original code)
2) Gcode preview 
3) SPI_GRAPHICAL_TFT option, because its doest work in my tft (at least by now)

So, in order to test, try disabling the two options I said above.

thisiskeithb commented:
> I think I know. The font load is broken. But I just dont changed it. I just disabled for my board. Disable it in your config file and test again, please.

That fixed the lock up, but the TFT is blank. The backlight is on, but tapping around doesn't result in broken icons like I was seeing before either. It's just blank.

rhapsodyv commented:
are you using SPI_GRAPHICAL_TFT ?

thisiskeithb commented:
> are you using SPI_GRAPHICAL_TFT ?

No, since it's not a Robin Nano 2.0 (I didn't have to enable it before, even if the icons were broken). I'm using `TFT_LITTLE_VGL_UI`.

rhapsodyv commented:
> > are you using SPI_GRAPHICAL_TFT ?
> 
> No, since it's not a Robin Nano 2.0 (I didn't have to enable it before, even if the icons were broken, the TFT was still "working"). I'm using `TFT_LITTLE_VGL_UI`.

Ok, I didnt test SPI_GRAPHICAL_TFT too.

Do you tried load the assets again? (the code renames the folder to bak_pic, to load again you need to rename back to mks_pic).

thisiskeithb commented:
> Did you try load the assets again? (the code renames the folder to bak_pic, to load again you need to rename back to mks_pic).

Yep. Each time I flash.

> are you using SPI_GRAPHICAL_TFT ?

I just enabled `SPI_GRAPHICAL_TFT` and the screen flashes white once on boot, then the backlight turns off/it doesn't work.

rhapsodyv commented:
try hardcode the DevideCode in Marlin/src/lcd/extui/lib/mks_ui/tft_lvgl_configuration.cpp.

change: 
```cpp
LCD_IO_WriteReg(0x00D3);
  DeviceCode = ILI9488_ReadRAM(); //dummy read
  DeviceCode = ILI9488_ReadRAM();
  DeviceCode = ILI9488_ReadRAM();
  DeviceCode <<= 8;
  DeviceCode |= ILI9488_ReadRAM();
```
to:
```cpp
LCD_IO_WriteReg(0x00D3);
  DeviceCode = ILI9488_ReadRAM(); //dummy read
  DeviceCode = ILI9488_ReadRAM();
  DeviceCode = ILI9488_ReadRAM();
  DeviceCode <<= 8;
  DeviceCode |= ILI9488_ReadRAM();

DeviceCode = 0x9488;
```

rhapsodyv commented:
> > What is the size your SPI Flash?
> 
> I'm not sure. Whatever is stock on the Robin Nano board/screen. It's an `STM32F103VET6`-based board.

The SPI Flash is something like W25Q80, or W25Q16... Mine is W25Q16. 16Mbit, 2MB....

rhapsodyv commented:
which board are you using? I will check the PINs... In my board I fixed all, but I didnt test others

thisiskeithb commented:
> What is the size your SPI Flash?

I think I found the chip. It's a [Winbond 25Q64JVSIQ](https://www.winbond.com/resource-files/w25q64jv%20spi%20%20%20revc%2006032016%20kms.pdf) 64M-bit / 8M-byte.

> which board are you using?

Robin Nano 1.2 with a "dumb" TFT35: [Direct link to pins file](https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/src/pins/stm32f1/pins_MKS_ROBIN_NANO.h).

rhapsodyv commented:
> try hardcode the DevideCode in Marlin/src/lcd/extui/lib/mks_ui/tft_lvgl_configuration.cpp.
> 
> change:
> 
> ```c++
> LCD_IO_WriteReg(0x00D3);
>   DeviceCode = ILI9488_ReadRAM(); //dummy read
>   DeviceCode = ILI9488_ReadRAM();
>   DeviceCode = ILI9488_ReadRAM();
>   DeviceCode <<= 8;
>   DeviceCode |= ILI9488_ReadRAM();
> ```
> 
> to:
> 
> ```c++
> LCD_IO_WriteReg(0x00D3);
>   DeviceCode = ILI9488_ReadRAM(); //dummy read
>   DeviceCode = ILI9488_ReadRAM();
>   DeviceCode = ILI9488_ReadRAM();
>   DeviceCode <<= 8;
>   DeviceCode |= ILI9488_ReadRAM();
> 
> DeviceCode = 0x9488;
> ```

Worked?

Other think I didnt test: their touch code. Their code for XPT2046 dint worked for me... So I just define TOUCH_BUTTONS to use marlin default XPT2046 implementation.

Try it too, in the config file:

```cpp
#define TOUCH_BUTTONS 
```

thisiskeithb commented:
Worked?

No. I get the "Booting.." text for a split second and then a blank screen.

>  Their code for XPT2046 dint worked for me

Touch worked fine on the Nano 1.2/TFT35 combo before this PR, but it was just the icons that didn't display correctly.

> #define TOUCH_BUTTONS 
> Maybe we can remove at all their touch code. I didnt because I dont have a MKS screen to test.

That'd be a no-go here since the "MKS UI" is meant as a replacement to all that. Like I mentioned above, touch worked fine on the Nano 1.2/TFT35 combo before this PR, but it was just the icons that didn't display correctly.

While this works fully on other hardware now, it no longer functions in any form on the Nano 1.2/TFT35.

thisiskeithb commented:
Flashing with [MKS's precompiled files](https://github.com/makerbase-mks/MKS-Robin-Nano/tree/master/Nano%20firmware%20v2.x.x) works, so that rules out hardware issues.

rhapsodyv commented:
> While this works fully on other hardware now, it no longer functions in any form on the Nano 1.2/TFT35.

It never worked in any hardware :-(  ... But give me some time, I will revise my changes. I may have removed some their code, or @thinkyhead did in the code cleanup.

rhapsodyv committed:
need to remove two extra pixel that dont exists

thisiskeithb commented:
> It never worked in any hardware :-( ... But give me some time

It did on my end, but the icons were not correct. Touch, menus, etc. worked, etc.

rhapsodyv commented:
> > It never worked in any hardware :-( ... But give me some time
> 
> It did on my end, but the icons were not correct. Touch, menus, etc. worked, etc.

What's your changes?

thisiskeithb commented:
> What's your changes?

None. I was using stock Marlin with `TFT_LITTLE_VGL_UI ` enabled: [Link to the config/build](https://github.com/thisiskeithb/Marlin/tree/testing/tevo-nereus-lvgl).

So like I said, while this PR appears to work on other hardware, the changes no longer work on the Nano/TFT35 now.

rhapsodyv commented:
> > What's your changes?
> 
> None. I was using stock Marlin with `TFT_LITTLE_VGL_UI ` enabled: [Link to the config/build](https://github.com/thisiskeithb/Marlin/tree/testing/tevo-nereus-lvgl).
> 
> So like I said, while this PR appears to work on other hardware, the changes no longer work on the Nano/TFT35 now.

:-O

How??

The original code didn't even start the TFT dma! Neither setup the touch pins!

Very weird. But I'm reviewing.

rhapsodyv commented:
I think is dma the issue. My board use dma. In the nano pins, it's dont use. I may have changed some line dependent upon dma.
I think I find.

rhapsodyv committed:
just call init if enabled

rhapsodyv committed:
reverting touch calibration to original values, and configuring it in the board; to not break the default MKS

rhapsodyv commented:
I broke the touch parameters. So I just fixed back to the original values, and made it configurable by board files - and changed only in my board.

thisiskeithb commented on Marlin/src/lcd/extui/lib/mks_ui/tft_lvgl_configuration.cpp:
```diff
@@ -434,7 +434,9 @@ void tft_lvgl_init() {
-  touch.init();
+  #if ENABLED(TOUCH_BUTTONS)
+    touch.init();
+  #endif
```
MKS' UI doesn't use `TOUCH_BUTTONS`, so should this if statement be here?

rhapsodyv replied:
MKS use the same SPI for flash and touch. I did this if so you can use the marlin way (separated spi) or the original MKS way (undef TOUCH_BUTTONS -> spi flash/touch together).

I just fixed back the touch I messed up

rhapsodyv replied:
It could be dying for you because this touch code init code... now with the #if, it wont blow up.

thisiskeithb replied:
There are two different build environments for the Robin Nano, so this isn't needed when building with MKS' LVGL.

You'd enable `TOUCH_BUTTONS` when using `FSMC_GRAPHICAL_TFT` for 12864/Emulated mode:

https://github.com/MarlinFirmware/Marlin/blob/8474050adbe1bd4d363397cfedefaee996ac1e98/platformio.ini#L543-L546

You'd use  `TFT_LITTLE_VGL_UI` (no `TOUCH_BUTTONS`) when building for the LVGL UI:

https://github.com/MarlinFirmware/Marlin/blob/8474050adbe1bd4d363397cfedefaee996ac1e98/platformio.ini#L630-L633

rhapsodyv replied:
Ok. In my case, I dont want the touch buttons on screen... The if is just to use the touch object (XPT2046 class).

After the last commit, black screen yet?

rhapsodyv replied:
The function: XPT2046_Rd_Addata only works if the board support have the touch and flash in the same SPI. MKS board is this way. Other boards dont. So I did this #if. 

So:
TFT_LITTLE_VGL_UI  with TOUCH_BUTTON  = separated SPI (my board)
TFT_LITTLE_VGL_UI without TOUCH_BUTTON  = same SPI (MKS board)
FSMC_GRAPHICAL_TFT with TOUCH_BUTTONS = any board that supports FSMC and touch

rhapsodyv commented:
I reviewed just touch and I think it back to the original behaviour. If we are using TOUCH_BUTTONS, it use separated SPI for touch. IF TOUCH_BUTTONS is disabled (default for MKS), the original MKS code runs.

rhapsodyv commented:
Nothing working yet?

I will get a nano to test. The only thing is different now is the initialization order, but I dont think it is the problem.

makerbase-mks commented:
@rhapsodyv Robin nano uses a hardware spi and shares a spi with spi_flash. When using lvgl, you do not need to enable the TOUCH_BUTTONS option

rhapsodyv commented:
Sure. I know.

But I made the code compatible with every board.

If the user enable touch buttons, the code will use the default marlin xpt class (my board)
If the user keep disable, it will use your original code.

This way, the lvgl ui will be compatible with more tft.
I’m currently using it in a chitu tft right now!!

A question: you don’t have a z offset screen right now?

thisiskeithb commented:
I pulled down the latest commits & applied the slightly modified settings to the configs from my [Tevo Nereus LVGL branch](https://github.com/thisiskeithb/Marlin/tree/testing/tevo-nereus-lvgl):

```cpp
#define HAS_SPI_FLASH_FONT 0
#define HAS_GCODE_PREVIEW 0
```

...and... Success!

![image](https://user-images.githubusercontent.com/13375512/85916251-b89e0d80-b803-11ea-80be-7c25a6eaa6e6.png)

![image](https://user-images.githubusercontent.com/13375512/85916243-aa4ff180-b803-11ea-97bb-5fdaa3c4ad28.png)

I can bug @makerbase-mks about the missing `bmp_Motor_off.bin` and `bmp_Print.bin` (which may have been renamed to `bmp_printing.bin`?), but the place holders should work for now.

SPI flash font is disabled with the flag above, so only English works. I do prefer the new sans serif font over serif font. Also, the menu is quite broken, probably related to above:
![image](https://user-images.githubusercontent.com/13375512/85916319-99ec4680-b804-11ea-84d0-29960f6d8699.png)

And of course the gcode preview is disabled, so it doesn't work.

> you don’t have a z offset screen right now?

I tapped around and didn't see it. Where should it be?

rhapsodyv commented:
Great news!!!! 

I forgot the icons of the language select screen!!

I will put it in the zip too.

But there’s plenty room to optimize the assets. The nano have a big spi flash, but I have only 2mb.
I will optimize this way:
- if is mks, I will load all Lang icons
- if is other board with small spi flash, I will disable screen Lang icons (keeps only text) or disable Lang select at all and use the build Lang

Gcode preview I will test tomorrow.

The most missing screen for me now is zoffset screen. It’s some of the most important screens to have.

thisiskeithb commented:
> The most missing screen for me now is zoffset screen. It’s some of the most important screens to have.

I don't think MKS' original TFT firmware had that screen unless it only shows up while printing (I'm just testing the Nano on a bench right now).

thisiskeithb commented:
@makerbase-mks: Do you have a newer version than what was supplied? It seems old since I'm missing several settings/features from [your video](https://www.youtube.com/watch?v=x03cVE1-gns) (and from your [precompiled firmware](https://github.com/makerbase-mks/MKS-Robin-Nano/tree/master/Nano%20firmware%20v2.x.x)):

Settings screenshot from the video:
![image](https://user-images.githubusercontent.com/13375512/85916670-ec7b3200-b807-11ea-8975-cb4c5cc721d8.png)

What I have:
![image](https://user-images.githubusercontent.com/13375512/85916689-103e7800-b808-11ea-88fd-af8b50c55a83.png)

makerbase-mks commented:
@rhapsodyv Yes, I haven't added this interface yet. I have modified some of the parameter adjustment interface, and I will submit it to the marlin team for review

thisiskeithb commented:
So we just have some outstanding issues below that @rhapsodyv will track down:

- If MKS Robin Nano, load all Lang icons
- Everything else/small spi flash, disable screen Lang icons (Use text only) or disable Lang select and use the build Lang

I'm not sure what to do about the missing fonts.

> Yes, I haven't added this interface yet. I have modified some of the parameter adjustment interface, and I will submit it to the marlin team for review

So that explains the missing code 🙂

makerbase-mks commented:
@thisiskeithb I have added some parameter adjustments, I think I will organize the code and submit it to you for review

thisiskeithb commented:
> I have added some parameter adjustments, I think I will organize the code and submit it to you for review

Submit code to @rhapsodyv's fork so it can be merged in with the working code in this PR: https://github.com/rhapsodyv/Marlin/tree/tft-little-vgl-ui-fixes-two

makerbase-mks commented:
@rhapsodyv This is the font file we use, hope it helps you.
https://github.com/makerbase-mks/Mks-Robin-Nano-Marlin2.0-Firmware/tree/master/Firmware/mks_font

thisiskeithb commented:
> @rhapsodyv This is the font file we use, hope it helps you.
> https://github.com/makerbase-mks/Mks-Robin-Nano-Marlin2.0-Firmware/tree/master/Firmware/mks_font

So that's why my icons weren't showing up prior to this PR, but the UI worked as it should.

I was using icons/fonts from https://github.com/makerbase-mks/MKS-Robin-Nano/tree/master/Nano%20firmware%20v2.x.x (old/different firmware) **and not the correct source of** https://github.com/makerbase-mks/Mks-Robin-Nano-Marlin2.0-Firmware/tree/master/Firmware. 

Having said that, if I start over with an LVGL build based on https://github.com/MarlinFirmware/Marlin/commit/8474050adbe1bd4d363397cfedefaee996ac1e98 and use the icons from makerbase-mks/Mks-Robin-Nano-Marlin2.0-Firmware, they are the "old" style (and fonts are broken, but I guess we knew that):

<p align="center"><img src="https://user-images.githubusercontent.com/13375512/85917651-d625a400-b810-11ea-9ee3-39b48e4db904.png" width="50%"></p>

makerbase-mks commented:
@thisiskeithb I just need to submit the code to @rhapsodyv's branch, right?

thisiskeithb commented:
> @thisiskeithb I just need to submit the code to @rhapsodyv's branch, right

For code fixes/updates, yes.

It’d make sense to host the images and fonts in Marlin as well for completeness, but that’s up to @thinkyhead.

makerbase-mks commented:
> @makerbase-mks: Will you upload the "new" 2.0 icons to your repo? @rhapsodyv has already converted most of them to work in his zip file.
Also, adding a comment above TFT_LITTLE_VGL_UI with a direct link to download the correct fonts and icons would be hugely beneficial so others don't make the same mistake slightly_smiling_face

In order to keep the icon names consistent, my current program has been changed to the original icon name. This is the 2.0 icon bin file I originally used
https://github.com/makerbase-mks/MKS-Robin-Nano/tree/master/Nano%20firmware%20v2.x.x

I don't know if this modification will conflict with the code of @rhapsodyv

rhapsodyv commented:
I think is better to have fixed names in code, and the files follow that.
This way, we will support any icon set, and the user may choose what they want.

The bg color can be a config too.

I will change to support the two icons set.

rhapsodyv commented:
I'm organising the asset load, to make it optimised to any SPI Flash size, loading only what is necessary according with the user options.

It will give us more flexibility and standardisation.

Preview (not finished yet):

<img width="324" alt="Screen Shot 2020-06-27 at 12 01 59" src="https://user-images.githubusercontent.com/81722/85925343-34cd3b00-b86e-11ea-8bde-8441012ee029.png">

rhapsodyv committed:
changed include order to fix redefinition warning

rhapsodyv committed:
the selected lang now is the same the user selected in the configuration.h

rhapsodyv committed:
using a icon that I could find

rhapsodyv committed:
Optimized asset load. Now the code just load the necessary assets according with the user options, so we can use this UI in a 2mb spi flash! Fix language screen; Change lookup to case insenstive

rhapsodyv commented:
- [x] Optimised asset load 
- [x] Enable/disable lang select screen.
- [x] Case insensitive lookup
- [ ] gcode preview
- [ ] font load/select (@makerbase-mks needs take a look in the font load code)

Future:
- [ ] option to change background, so the user can use more than one icon set
- [ ] zoffset / baby steps screen 

Updated assets folder:
https://drive.google.com/file/d/1CSCi-306uCIV7Yz9wasTV3_5iZpTSmIO/view?usp=sharing

(I didnt test yet enabling language select screen, I will need to remove some icons to fit in my flash to test)

rhapsodyv committed:
index issue

rhapsodyv committed:
align

rhapsodyv committed:
icons for screen that only shows while printing

rhapsodyv commented:
@makerbase-mks Who can I generate I gcode preview in cura? I need it to test the code.

rhapsodyv committed:
fixes file image preview in the file selection screen

rhapsodyv commented:
Just fixed the image file preview... 

Steps: I downloaded the MKS Plugin for cura and saved the gcode with preview.

Before the last commit:
![IMG_9349](https://user-images.githubusercontent.com/81722/85935804-f7e36180-b8ca-11ea-916d-fb35fd23c2bb.jpg)

Now:

![IMG_9350](https://user-images.githubusercontent.com/81722/85935814-029df680-b8cb-11ea-83ef-f8d785734791.jpg)

rhapsodyv committed:
now it works with the 50x50 and 100x100 icons!

rhapsodyv commented:
The current MKS firmware expect 100x100 icons, but the MKS Plugin are sending 50x50 icons. Some users have changed the plugin to 100x100. 
So, I made the code works with both sizes!!

![IMG_9351](https://user-images.githubusercontent.com/81722/85936575-9626f580-b8d2-11ea-9f28-813df3908b8a.jpg)

thinkyhead commented:
Thanks for making checkboxes. We should use those more.

rhapsodyv committed:
gcode preview fully working

rhapsodyv commented:
Preview fully working, even with the small images generated by the MKS Cura Plugin

![IMG_9354](https://user-images.githubusercontent.com/81722/85941159-81108d80-b8f7-11ea-98db-4abb2e6fa0b1.jpg)
![IMG_9355](https://user-images.githubusercontent.com/81722/85941160-8372e780-b8f7-11ea-8a39-da532142dd6d.jpg)

rhapsodyv commented:
I think the font issue isn’t necessary for now, because the code are using the lvgl default font.

In my option, the most important missing feature is a zoffset screen, because it’s very needed by the end user.

Anyway, all the other stuff are working now. Better yet, it’s working not only in mks board and tft, it will work in any tft with at least 2mb spi flash!!!  (Like my tronxy x5sa!)

thisiskeithb commented:
> I think the font issue isn’t necessary for now, because the code are using the lvgl default font.

I’m not near a computer right now, but does multi-language work?

Also, should the (now fixed) default MKS icons be included in Marlin?

rhapsodyv commented:
> Also, should the (now fixed) default MKS icons be included in Marlin?

We can have more than one icon set / theme. The zip file I sent is a copy of MKS icons, but just the necessary by now.

> I’m not near a computer right now, but does multi-language work?
Yes. Take a look:
![ezgif-4-2223c183518f](https://user-images.githubusercontent.com/81722/85953833-93b3b280-b949-11ea-88c8-3f6c8e94a34a.gif)

rhapsodyv commented:
@makerbase-mks @thinkyhead @thisiskeithb 

Here is a compiled list of what I did.

The main goal were: 
1. Integrate better with current Marlin code
2. Fix some issues
3. Make it more generic so more boards can use it

### Asset management

- Fixed a issue that prevent load if one of the files fails to read long names (UpdatePic loop)
- Fixed the names of some icons (motor_off, print, and a few others)
- Changed the lookup to be case insensitive, to avoid naming problems.
- Created a folder with just the necessary icons
- Fixed bg color
- Font loading is optional. If no font are loaded, the code will use the LVGL default font (roboto 22)
- Created flags for diferente assets needs. User can enable/disable: Font loading, gcode preview, preview backup in flash, language selection screen. It allows the next fix: optimised asset loading.
- Optimised the asset loading. Now the code only loads the icons that match the user compilation flags. It saves a lot of spi space.

Now we have a list of assets that need to load according with the user flags. It allowed me to fit all assets in a 2MB spi flash!! 

### Language

- Lots translations with _T_ in their name was missing. I changed them to a non _T_ version and it worked.
- The default lang now is the same of the selected in Configuration.h for the user.
- The user can disable the language selection screen to save spi flash.

### Gcode preview

- The old code works only if the small image is 100x100. But, the current MKS plugin for cura send images in 50x50. Some users have fixed the cura plugin, others dont. Even the last version is using 50x50 yet. So we have users with two different small image sizes. I made the code working with *both* sizes.
- User can disable gcode preview and backup to save flash

### Touch and TFT

The old code works just if the board share the SPI for touch and flash and have no dma in TFT. To more boards use this UI, a made:

- Made the TFT works if the user have a DMA enabled TFT
- If the user define TOUCH_BUTTONS, it allows the code to use different SPI for flash and touch.
- Fixed the tft initialisation order (it tried use it before init..)
- Made it working in Chitu Board and Stock TFT!

### Things I could not test

- SPI_GRAPHIC_TFT paths
- Font asset loading / usage (the code loads only one file, but mks have 2). 



I think that's it.

thisiskeithb commented:
Bravo getting this into a useable state. Hopefully MKS can work out the remaining issues & bring the LVGL-based firmware in line with their old/proprietary TFT firmware.

> Created a folder with just the necessary icons

Is [this](https://drive.google.com/file/d/1CSCi-306uCIV7Yz9wasTV3_5iZpTSmIO/view?usp=sharing) the most up to date source? If so, it'd be good if @makerbase-mks could update their repo with it and a link be added near `TFT_LITTLE_VGL_UI` in `Configuration.h` so users can easily find the "default" icons.

rhapsodyv commented:
> Bravo getting this into a useable state. Hopefully MKS can work out the remaining issues & bring the LVGL-based firmware in line with their old/proprietary TFT firmware.
> 
> > Created a folder with just the necessary icons
> 
> Is [this](https://drive.google.com/file/d/1CSCi-306uCIV7Yz9wasTV3_5iZpTSmIO/view?usp=sharing) the most up to date source? If so, it'd be good if @makerbase-mks could update their repo with it and a link be added near `TFT_LITTLE_VGL_UI` in `Configuration.h` so users can easily find the "default" icons.

I changed some icon names to match de code... But, maybe is better to take their original mks_pic folder and change the code to match the icons... 

Good Idea about the Configuration.h.

thinkyhead commented:
I'm working on 2.0.6 release, and thinking it will be at around midnight tonight CST, so in about 8 hours.

If this is in "beta" shape by that time then it can be included and marked as such.

rhapsodyv commented:
> I'm working on 2.0.6 release, and thinking it will be at around midnight tonight CST, so in about 8 hours.
> 
> If this is in "beta" shape by that time then it can be included and marked as such.

For chitu board its working fine. @thisiskeithb can you test it on your mks nano?

I will do 2 things now for the release:

1. Change the icon names in the code to match the MKS supplied zip folder: https://github.com/makerbase-mks/MKS-Robin-Nano/tree/master/Nano%20firmware%20v2.x.x

2. Change the Configuration.h, pointing their icon URL (above), as suggested by @thisiskeithb . So the user will know what to do when activate the TFT_LITTLE_VGL_UI.

OR, Marlin can host the icon set... But I think by now it's better to point to MKS Repo.

thisiskeithb commented:
> Change the Configuration.h, pointing their icon URL (above), as suggested by @thisiskeithb . So the user will know what to do when activate the TFT_LITTLE_VGL_UI.

Something like this would work (shortened link points to MKS' repo):

```cpp
//
// TFT Little VGL UI
// Default MKS icons and fonts: https://git.io/JJvzK
//
//#define TFT_LITTLE_VGL_UI
```

@rhapsodyv: Will you put in a PR to MKS' repo with the 2.0-style icons once renamed?

rhapsodyv commented:
> > Change the Configuration.h, pointing their icon URL (above), as suggested by @thisiskeithb . So the user will know what to do when activate the TFT_LITTLE_VGL_UI.
> 
> Something like this would work (shortened link points to MKS' repo):
> 
> ```c++
> //
> // TFT Little VGL UI
> // Default MKS icons and fonts: https://git.io/JJvzK
> //
> //#define TFT_LITTLE_VGL_UI
> ```
> 
> @rhapsodyv: Will you put in a PR to MKS' repo with the 2.0-style icons once renamed?
> 
> > can you test it on your mks nano?
> 
> Text only shows up on the Nano if `#define HAS_SPI_FLASH_FONT 0` is in the config/pins, so that should probably be added to the pins file under the `#elif ENABLED(TFT_LITTLE_VGL_UI)` block until loading external fonts works.
> 
> Other than that, this is in a totally workable/beta state 🙂

Ok. I will edit picmanger to not try load custom fonts by default, until we fix that. So we dont need edit the board.

I will edit the config too.

Here is the differences between code and their zip that I will rename:

In Code, but not in their ZIP | Zip "equivalent"
-- | --
bmp_auto.bin | bmp_auto_off.bin
bmp_motor_off.bin | bmp_manual_off.bin
bmp_extruct_speed.bin | bmp_extrude_opr.bin
bmp_extruct_speed_sel.bin | bmp_extruct_sel.bin
bmp_mamual.bin | bmp_manual_off.bin
bmp_mov_speed.bin | bmp_mov_changespeed.bin
bmp_mov_speed_sel.bin | bmp_mov_sel.bin
bmp_print.bin | bmp_printing.bin
bmp_simple_cn.bin | bmp_simplified_cn.bin
bmp_simple_cn_sel.bin | bmp_simplified_cn_sel.bin
bmp_tradition_cn.bin | bmp_traditional_cn.bin
bmp_tradition_cn_sel.bin | bmp_traditional_cn_sel.bin

thisiskeithb commented:
I edited my comment:

> Also, there's no Language screen. Should there be a define for that as well or is that related to the external fonts not working?

thisiskeithb commented:
@thinkyhead What are your thoughts on including MKS' fixed icons/fonts with Marlin? They can be stored under Marlin/src/lcd/extui/lib/mks_ui/assets (wherever else would make sense).

Then `Configuration.h` could be updated:

```cpp
//
// TFT Little VGL UI
// Copy mks_pic and mks_font folders from Marlin/src/lcd/extui/lib/mks_ui/assets to the root of your SD card
//
//#define TFT_LITTLE_VGL_UI
```

rhapsodyv committed:
some files dont have the LONG_FILE_NAME flag, but are valid and needs to work

rhapsodyv committed:
fixed icon namming to match with MKS icons release

rhapsodyv committed:
fixes sizes and options for chitu board, to fit the current assets

rhapsodyv commented:
Now it will work with MKS icons folder.

@thisiskeithb can you test again?

For Nano Board, the defaults are: font load off, gcode preview on, lang selection on (menu settings).

rhapsodyv committed:
missing include to HAS_LANG_SELECT_SCREEN work

thisiskeithb commented:
With your last push, the language screen now shows up. 🙂 We can't do much with a few languages since external fonts/the default LVGL font doesn't appear to support unicode, but English, Spanish, French, and Italian languages work.

<p align="center"><img src="https://user-images.githubusercontent.com/13375512/85967775-1a7f8400-b978-11ea-8b1a-85ca7cce1c4f.png" width="50%"></p>

Also, MKS updated their icons while we were testing this PR, so they are the new "2.0" style. Yay!

This is ready for merging as a beta feature as far as I'm concerned.

Great work @rhapsodyv!

rhapsodyv committed:
disable lang selection screen, because the lvgl current default font only support english letters

rhapsodyv committed:
keep only english, until we have a font that support more characters

thisiskeithb commented:
I had a look at the Language screen again and noticed accented letters were missing from the default LVGL font (ex: Espaol should be Español), so @rhapsodyv disabled language switching for now.

@makerbase-mks: Updating to LVGL 7 & getting fonts to load should be on your to-do list for better language support with accented characters 🙂

GrumpaV2 commented:
Victor did you test on Robin_Nano 1.2?

rhapsodyv commented:
> Victor did you test on Robin_Nano 1.2?

@thisiskeithb did. Why? Did you find any problem?

rhapsodyv commented:
@makerbase-mks I think the "base" is pretty stable now. We can work together to bring more screens to Marlin.

thisiskeithb commented:
This shouldn’t hold up the PR, but the project is now legally/officially “LVGL” which stands for “Light and Versatile Graphics Library” [as of 7.0](https://github.com/lvgl/lvgl/releases/tag/v7.0.0).

They must have ran into a legal issue with the former “Little” name.

makerbase-mks commented:
@rhapsodyv Can I test your modified version? I think I also need to make some changes.

> This shouldn’t hold up the PR, but the project is now legally/officially “LVGL” which stands for “Light and Versatile Graphics Library” as of 7.0.

They must have ran into a legal issue with the former “Little” name.

@thisiskeithb In order to legalize the use of LVGL, do we need to upgrade LVGL to version 7.0?

makerbase-mks commented:
https://github.com/makerbase-mks/Mks-Robin-Nano-Marlin2.0-Firmware/tree/master/Firmware
I have upgraded the icon to 2.0, and amended the corresponding icon name, I think I will continue to use this general format in the future

thisiskeithb commented:
> @thisiskeithb In order to legalize the use of LVGL, do we need to upgrade LVGL to version 7.0?

I don’t think so. It was more of a heads up for @thinkyhead that the name changed since were still using “little” in `TFT_LITTLE_VGL_UI`.

But moving to 7.0 wouldn’t hurt since it has better font support 🙂

rhapsodyv commented:
> @rhapsodyv Can I test your modified version? I think I also need to make some changes.

Sure. I can help you to merge the changes back to your repo. I'm (and a lot of users) anxious for more screens using LVGL :-). 
This UI is very beautiful and user friendly!

I'm at your disposal.


About the LVGL 7. I don't what everything you changed in your fork, but one think the 7 version will solve: the image load for buttons. Now they have a way to create loaders, so we dont need change their core for that.
And we can test their updated font support too.

rhapsodyv commented:
> > @thisiskeithb In order to legalize the use of LVGL, do we need to upgrade LVGL to version 7.0?
> 
> I don’t think so. It was more of a heads up for @thinkyhead that the name changed since were still using “little” in `TFT_LITTLE_VGL_UI`.
> 
> But moving to 7.0 wouldn’t hurt since it has better font support 🙂

We can rename it to `TFT_LVGL_UI`... right now?

thisiskeithb commented:
With `2.0.6` being so close to release, I wasn’t sure if it’d be good to do it now or wait.

rhapsodyv commented:
> With `2.0.6` being so close to release, I wasn’t sure if it’d be good to do it now or wait.

It's just a search and replace. But I know what you mean. As it will be a "beta" feature, I dont think it's a problem change it later. I don't even know when we will or not migrate to version 7.

MKS-Sean commented:
![image](https://user-images.githubusercontent.com/56996910/86090974-b080ef00-badd-11ea-958a-f97d1d30da9b.png)
I tested in nano 1.2 to display fonts. Is the problem of not displaying fonts on other motherboards SPI_FLASH not enough space?

makerbase-mks commented:
> About the LVGL 7. I don't what everything you changed in your fork, but one think the 7 version will solve: the image load for buttons. Now they have a way to create loaders, so we dont need change their core for that.
And we can test their updated font support too.

The preview icon display and the button icon display require us to change the LVGL code. I don’t know what the 7.0 version has changed. I will test it, but I think I should synchronize the current code with you first. I will Test your version and submit it to you after merging

thisiskeithb commented:
> I tested in nano 1.2 to display fonts. Is the problem of not displaying fonts on other motherboards SPI_FLASH not enough space?

I was testing this code on a Nano 1.2 as well and non-English fonts weren’t displaying correctly. Chinese and Russian weren’t even showing up, so how’d you fix that?

<p align="center"><img src="https://user-images.githubusercontent.com/13375512/85967775-1a7f8400-b978-11ea-8b1a-85ca7cce1c4f.png" width="50%"></p>

rhapsodyv commented:
> > About the LVGL 7. I don't what everything you changed in your fork, but one think the 7 version will solve: the image load for buttons. Now they have a way to create loaders, so we dont need change their core for that.
> > And we can test their updated font support too.
> 
> The preview icon display and the button icon display require us to change the LVGL code. I don’t know what the 7.0 version has changed. I will test it, but I think I should synchronize the current code with you first. I will Test your version and submit it to you after merging

I dont know in v6, but in 7 we can create a image_decoder that is responsible to load any image by their name, take a look: https://docs.lvgl.io/latest/en/html/overview/image.html#register-an-image-decoder

So we can just use: `lv_img_set_src(icon, "S:my_icon.bin");`, and the decoder will be trigged to load the image. There're more options too, like a "file system loader", etc.

rhapsodyv commented:
> ![image](https://user-images.githubusercontent.com/56996910/86090974-b080ef00-badd-11ea-958a-f97d1d30da9b.png)
> I tested in nano 1.2 to display fonts. Is the problem of not displaying fonts on other motherboards SPI_FLASH not enough space?

The problem with the font loader was: I could not make it work. In the mks_font I are able to find, it always had two files: GBK16.bin and UNIGBK.bin, but the code loads only one. And none of them worked. 

I didnt test with the last updated you did in your repository. I saw that now you have only one file: FontUNIGBK.bin ... I will test if it work.

rhapsodyv commented:
@thisiskeithb can you test with the new font?

just enable:
```cpp
#define HAS_SPI_FLASH_FONT 1
#define HAS_LANG_SELECT_SCREEN 1
```

And use the this font:  https://github.com/makerbase-mks/Mks-Robin-Nano-Marlin2.0-Firmware/tree/master/Firmware/mks_font

thisiskeithb commented:
> just enable:
> 
> ```c++
> #define HAS_SPI_FLASH_FONT 1
> #define HAS_LANG_SELECT_SCREEN 1
> ```
> 
> And use the this font: https://github.com/makerbase-mks/Mks-Robin-Nano-Marlin2.0-Firmware/tree/master/Firmware/mks_font

I enabled those options & used the updated `FontUNIGBK.bin` from the link above, but now the text is missing below the icons in all the screens.

rhapsodyv commented:
> > just enable:
> > ```c++
> > #define HAS_SPI_FLASH_FONT 1
> > #define HAS_LANG_SELECT_SCREEN 1
> > ```
> > 
> > 
> > And use the this font: https://github.com/makerbase-mks/Mks-Robin-Nano-Marlin2.0-Firmware/tree/master/Firmware/mks_font
> 
> I recompiled the firmware from this PR with those options enabled & used the updated `FontUNIGBK.bin` & icons from the link above, but now the text is missing below the icons in all the screens.
> 
> Here's a video of the update process and as you can see, the main binary, images, and font are being flashed: https://youtu.be/5xmuoIucRDs
> 
> [Here's a link](https://github.com/thisiskeithb/Marlin/tree/testing/tevo-nereus-lvgl) to my Tevo Nereus config (Robin Nano 1.2/TFT35) rebased on top of this PR for testing.

@makerbase-mks I think the problem is with the offsets/address you are using. To test in my board I had to change my offsets to fit in a 2mb flash. And it worked for my (buggy - note the dots bellow the font, but the font was loaded) (ignore the missing icons, I had to unload them to load the font...)
Looks like you need revise your font addresses/offsets. I can't test it right now, because the size of my flash.
I brought a nano board, but it will arrive just in a few weeks.

![IMG_9386](https://user-images.githubusercontent.com/81722/86140168-f67e8880-bac6-11ea-92d1-29b42d6c0f15.jpg)

thisiskeithb commented:
I created issue https://github.com/MarlinFirmware/Marlin/issues/18476 so discussion can continue over there and this PR can be merged as-is with the assumption it's still in beta.

Since the last couple commits in this PR default to the built-in font & English only without the language screen, it's good to go for now. We can work out the rest after it's merged.

thinkyhead committed:
Update name to LVGL
